File: src/ros_apps/carla_telemetry/carla_telemetry/sensors/imu.py
"""
| File: imu.py
| Description: CARLA IMU sensor wrapper.
               Wraps sensor.other.imu and collects accelerometer, gyroscope,
               and compass readings for publishing as sensor_msgs/Imu.

CARLA IMU data members:
  data.accelerometer : carla.Vector3D  (m/s², CARLA frame: X=fwd, Y=right, Z=up)
  data.gyroscope     : carla.Vector3D  (rad/s)
  data.compass       : float           (degrees, 0=North, clockwise positive)
  data.transform     : carla.Transform (sensor world pose)
"""
__all__ = ["CarlaIMU"]

import math
import threading
import weakref


class CarlaIMU:
    """CARLA IMU sensor wrapper.

    Output state dict keys:
    ===================== =====================================================
    Key                   Value
    ===================== =====================================================
    ``accel_x/y/z``       Linear acceleration in CARLA body frame (m/s²)
    ``gyro_x/y/z``        Angular velocity (rad/s)
    ``compass``           Compass heading (degrees, 0=North CW)
    ``frame_id``          ROS 2 frame ID string
    ===================== =====================================================
    """

    def __init__(self, carla_world, vehicle_actor, config: dict):
        """
        Args:
            carla_world:   carla.World instance.
            vehicle_actor: Ego vehicle actor.
            config:        IMU config dict from carla_interface_config.yaml.
        """
        self._frame_id = str(config.get("frame_id", "imu_link"))
        self._lock = threading.Lock()
        self._last_data = None

        # Spawn blueprint
        import carla
        bp = carla_world.get_blueprint_library().find("sensor.other.imu")

        noise_attrs = [
            "noise_accel_stddev_x", "noise_accel_stddev_y", "noise_accel_stddev_z",
            "noise_gyro_stddev_x",  "noise_gyro_stddev_y",  "noise_gyro_stddev_z",
            "noise_gyro_bias_x",    "noise_gyro_bias_y",    "noise_gyro_bias_z",
        ]
        for attr in noise_attrs:
            val = config.get(attr)
            if val is not None and bp.has_attribute(attr):
                bp.set_attribute(attr, str(float(val)))

        sp_cfg = config.get("spawn_point", {})
        transform = carla.Transform(
            carla.Location(
                x=float(sp_cfg.get("x", 0.0)),
                y=float(sp_cfg.get("y", 0.0)),
                z=float(sp_cfg.get("z", 0.0)),
            ),
            carla.Rotation(
                roll=float(sp_cfg.get("roll", 0.0)),
                pitch=float(sp_cfg.get("pitch", 0.0)),
                yaw=float(sp_cfg.get("yaw", 0.0)),
            ),
        )
        self._sensor = carla_world.spawn_actor(bp, transform, attach_to=vehicle_actor)

        weak_self = weakref.ref(self)
        self._sensor.listen(lambda data: CarlaIMU._on_imu(weak_self, data))

    # ------------------------------------------------------------------
    # CARLA callback
    # ------------------------------------------------------------------

    @staticmethod
    def _on_imu(weak_self, data):
        self = weak_self()
        if self is None:
            return
        with self._lock:
            self._last_data = data

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get_state(self):
        """Return latest IMU state dict or None if no data yet."""
        with self._lock:
            data = self._last_data

        if data is None:
            return None

        limits = (-99.9, 99.9)

        def clamp(v):
            return max(limits[0], min(limits[1], v))

        return {
            "accel_x": clamp(data.accelerometer.x),
            "accel_y": clamp(data.accelerometer.y),
            "accel_z": clamp(data.accelerometer.z),
            "gyro_x":  clamp(data.gyroscope.x),
            "gyro_y":  clamp(data.gyroscope.y),
            "gyro_z":  clamp(data.gyroscope.z),
            "compass": math.degrees(data.compass) if hasattr(data.compass, '__float__') else float(data.compass),
            "frame_id": self._frame_id,
        }

    def destroy(self):
        """Stop and destroy the CARLA IMU sensor."""
        if self._sensor is not None:
            try:
                self._sensor.stop()
                self._sensor.destroy()
            except Exception:
                pass
            self._sensor = None

File: src/ros_apps/carla_telemetry_cpp/src/sensors/imu.cpp
#include "carla_telemetry/sensors/imu.hpp"

#include <carla/actors/BlueprintLibrary.h>
#include <carla/geom/Transform.h>
#include <tf2/LinearMath/Quaternion.h>

#include <algorithm>
#include <cmath>

#include "carla_telemetry/perf_monitor.hpp"

namespace carla_telemetry {

CarlaIMU::CarlaIMU(carla::client::World& world,
                   carla::SharedPtr<carla::client::Actor> vehicle,
                   const std::string& frame_id, float x, float y, float z,
                   float roll, float pitch, float yaw)
    : frame_id_(frame_id) {
  auto bp_lib = world.GetBlueprintLibrary();
  auto bp = *bp_lib->Find("sensor.other.imu");

  carla::geom::Transform transform(carla::geom::Location(x, y, z),
                                   carla::geom::Rotation(pitch, yaw, roll));

  auto actor = world.SpawnActor(bp, transform, vehicle.get());
  sensor_ = boost::dynamic_pointer_cast<carla::client::Sensor>(actor);

  sensor_->Listen([this](auto data) { this->on_imu(data); });
}

void CarlaIMU::on_imu(carla::SharedPtr<carla::sensor::SensorData> data) {
  auto imu =
      boost::dynamic_pointer_cast<carla::sensor::data::IMUMeasurement>(data);
  if (!imu) return;
  // Rate the server delivers IMU measurements to this client (inter-arrival).
  perf.record_interval(frame_id_ + ".server_dt");
  std::lock_guard<std::mutex> lk(lock_);
  last_data_ = imu;
}

std::optional<ImuState> CarlaIMU::get_state() {
  carla::SharedPtr<carla::sensor::data::IMUMeasurement> data;
  {
    std::lock_guard<std::mutex> lk(lock_);
    data = last_data_;
  }
  if (!data) return std::nullopt;

  auto clamp = [](double v) -> double { return std::clamp(v, -99.9, 99.9); };

  auto accel = data->GetAccelerometer();
  auto gyro = data->GetGyroscope();
  auto transform = data->GetSensorTransform();
  auto& rot = transform.rotation;

  double roll = static_cast<double>(rot.roll) * M_PI / 180.0;
  double pitch = static_cast<double>(-rot.pitch) * M_PI / 180.0;
  double yaw = M_PI / 2.0 - static_cast<double>(data->GetCompass());

  tf2::Quaternion q;
  q.setRPY(roll, pitch, yaw);  //+ 0.2098132

  ImuState s;
  s.accel_x = clamp(static_cast<double>(accel.x));
  s.accel_y = clamp(static_cast<double>(accel.y));
  s.accel_z = clamp(static_cast<double>(accel.z));
  s.gyro_x = clamp(static_cast<double>(gyro.x));
  s.gyro_y = clamp(static_cast<double>(gyro.y));
  s.gyro_z = clamp(static_cast<double>(gyro.z));
  s.qx = q.x();
  s.qy = q.y();
  s.qz = q.z();
  s.qw = q.w();
  s.frame_id = frame_id_;
  return s;
}

void CarlaIMU::destroy() {
  if (sensor_) {
    try {
      sensor_->Stop();
      sensor_->Destroy();
    } catch (...) {
    }
    sensor_ = nullptr;
  }
}

}  // namespace carla_telemetry
